src/main/scala/org/constellation/domain/redownload/RedownloadService.scala

```diff
@@ -1,38 +1,159 @@
+import org.constellation.p2p.Cluster
 import org.constellation.schema.Id
+import org.constellation.serializer.KryoSerializer
+import org.constellation.storage.RecentSnapshot
+import org.constellation.util.{HealthChecker, MajorityStateChooser, SnapshotDiff}
 
-class RedownloadService[F[_]]()(implicit F: Concurrent[F]) {
+import scala.concurrent.duration._
+
+class RedownloadService[F[_]](cluster: Cluster[F], healthChecker: HealthChecker[F])(implicit F: Concurrent[F],
+                                                                                    C: ContextShift[F]) {
+  type Proposals = Map[Id, RecentSnapshot]
+  val logger = Slf4jLogger.getLogger[F]
 
   // TODO: Consider Height/Hash type classes
-  private[redownload] val ownSnapshots: Ref[F, Map[Long, String]] = Ref.unsafe(Map.empty)
-  private[redownload] val peersProposals: Ref[F, Map[Long, (Id, String)]] = Ref.unsafe(Map.empty)
+  private[redownload] val proposedSnapshots: Ref[F, Map[Long, Proposals]] = Ref.unsafe(Map.empty)
 
-  def persistOwnSnapshot(height: Long, hash: String): F[Unit] =
-    ownSnapshots.modify { m =>
-      val updated = if (m.get(height).nonEmpty) m else m.updated(height, hash)
-      (updated, ())
-    }
+  def persistLocalSnapshot(recentSnapshot: RecentSnapshot) = updateProposedSnapshots(Seq((cluster.id, recentSnapshot)))
 
-  def getOwnSnapshots(): F[Map[Long, String]] = ownSnapshots.get
+  def getLocalSnapshots() = proposedSnapshots.get.map { snapshots =>
+    snapshots.values.flatMap { proposalsForHeight =>
+      proposalsForHeight.get(cluster.id)
+    }.toSeq
+  }
 
-  def getOwnSnapshot(height: Long): F[Option[String]] =
-    ownSnapshots.get.map(_.get(height))
+  def getLocalSnapshotAtHeight(height: Long): F[Option[RecentSnapshot]] = proposedSnapshots.get.map { proposedSnaps =>
+    val snapshotsAtHeight = proposedSnaps.getOrElse(height, Map())
+    snapshotsAtHeight.get(cluster.id)
+  }
 
-  // TODO: Use cluster peers to fetch "/snapshot/own" and merge with peersProposals Map
-  def fetchPeersProposals(): F[Unit] = ???
+  private[redownload] def fetchPeerProposals(): F[List[(Id, RecentSnapshot)]] =
+    for {
+      peers <- cluster.readyPeers
+      chunkedPeerProposals <- peers.toList.traverse {
+        case (id, peerData) =>
+          val resp = peerData.client.getNonBlockingFLogged[F, Array[Array[Byte]]](
+            "snapshot/own",
+            timeout = 3.seconds,
+            tag = "snapshot/own"
+          )(C)
+          resp.map((id, _)).handleErrorWith { ex =>
+            Sync[F].delay(logger.error(s"fetchPeerProposals error - ${ex}")) >> Sync[F].delay((id, Array()))
+          }
+      }
+      deSerializedPeerProposals = chunkedPeerProposals.map(RedownloadService.deserializeProposals).flatMap {
+        case (id, recentSnaps) => recentSnaps.map(snap => (id, snap))
+      }
+    } yield deSerializedPeerProposals
+
+  private[redownload] def updateProposedSnapshots(fetchedProposals: Seq[(Id, RecentSnapshot)]) =
```

Comment:
why it takes `Seq` as an argument? it should be a map `Map[Id, RecentSnapshot]` or two separated arguments, so `(id: Id, snapshot: RecentSnapshot)`, depending on the use case.